smithy-build/src/main/java/software/amazon/smithy/build/ProtoReservedFieldsTrait.java

```diff
@@ -0,0 +1,63 @@
+package software.amazon.smithy.build;
+
+import java.util.List;
+import java.util.ArrayList;
+
+import software.amazon.smithy.model.*;
+import software.amazon.smithy.model.node.*;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.*;
+import software.amazon.smithy.utils.ListUtils;
+
+public final class ProtoReservedFieldsTrait extends AbstractTrait {
+	public static final ShapeId ID = ShapeId.from("ns.foo#protoReservedFields");
+
+	private final List<Integer> reserved;
+
+	public ProtoReservedFieldsTrait(Builder builder) {
+		super(ID, SourceLocation.NONE);
+		this.reserved = ListUtils.copyOf(builder.reserved);
+	}
+
+	public List<Integer> getReserved() {
+		return this.reserved;
+	}
+
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	public static final class Builder extends AbstractTraitBuilder<ProtoReservedFieldsTrait, Builder> {
+		private final List<Integer> reserved = new ArrayList<>();
+
+		public Builder add(Integer reserved) {
+			this.reserved.add(reserved);
+			return this;
+		}
+
+		@Override
+		public ProtoReservedFieldsTrait build() {
+			return new ProtoReservedFieldsTrait(this);
+		}
+	}
+
+	public static final class Provider extends AbstractTrait.Provider {
+		public Provider() {
+			super(ID);
+		}
+
+		@Override
+		public ProtoReservedFieldsTrait createTrait(ShapeId target, Node value) {
+			Builder builder = new Builder().sourceLocation(value);
+			for (Node definition : value.expectArrayNode().getElements()) {
+				definition.asNumberNode().map(NumberNode::getValue).map(Number::intValue).ifPresent(builder::add);
+			}
+			return builder.build();
+		}
+	}
+
+	@Override
+	protected Node createNode() {
+		return Node.arrayNode();
```

Comment:
ahh it would be the implementation on my side, let me give it a shot

Reply:
wow it's my mistake, I'm sorry for wasting your time. I assumed this code was fine (did not write it) and only found an issue when I tried to write a transformer...

correct implementation would look like:
```
	protected Node createNode() {
		return reserved.stream().map(number -> new NumberNode(number, getSourceLocation())).collect(ArrayNode.collect(getSourceLocation()));
	}
```

Reply:
Yeah in general, the toNode implementation of a trait is super important (e.g., used for equals and hashCode), but in the case of RenameShapes, it's a hard requirement IIRC because the model is serialized then reserialized to handle the renaming.